TeensyDmx.cpp

```diff
@@ -892,7 +994,514 @@ void TeensyDmx::maybeIncrementChecksumFail()
     }
 }
 
-void TeensyDmx::processRDM()
+
+void TeensyDmx::doRDMDiscovery() {
+    m_uidCount = 0;
+    m_dubPointer = 0;
+    m_discoveryState = DiscoveryState::DISCOVERY_UN_MUTE;
+    // Action queued, do it ASAP
+    m_nextDiscoveryAction = 0;
+    m_dubLowerBoundUid = RDM_MIN_LOWER_BOUND_UID;
+    m_dubUpperBoundUid = RDM_MAX_UPPER_BOUND_UID;
+    m_dubQueue[(m_dubPointer * 2)] = m_dubLowerBoundUid;
+    m_dubQueue[(m_dubPointer * 2) + 1] = m_dubUpperBoundUid;
+    m_dubPointer++;
+}
+
+void TeensyDmx::maybeTimeoutRDMMessage() {
+    if (m_controllerState != ControllerState::CONTROLLER_IDLE) {
+        if (millis() >= m_rdmResponseDue) {
+            switch (m_controllerState)
+            {
+                case ControllerState::RDM_DUB:
+                    if (m_state != State::RDM_DUB_PRE_PREAMBLE) {
+                      // A timed out DUB with data becomes a collision
+                      m_controllerState = ControllerState::RDM_DUB_COLLISION;
+                    } else {
+                      // A timed out DUB with no data becomes a timeout
+                      Serial.print("DUB timing out, RDM: ");
```

Comment:
Debugging code in library.  Presumably this is something you're still working on?

Reply:
Yes and no. There's a whole pile of corner cases for discovery, for well behaved responders, let alone buggy ones, so I'd like to try and come to some sort of compromise on this. Ideally with a macro or something clever, so a developer in the field can turn on some suitable debugging relatively easily, but I'd be willing to just having to sed uncomment a load of prints or something to do so.

I guess the ideal would be some form of logging implementation, with different levels where the developer using the library can choose what they log, and to where, but I don't know if that kind of thing exists in uC land.